transifex/openformats
# Introduce handler for .docx files

Problem and/or solution
-----------------------

With this commit a new handler for handling .docx files
is introduced. A high level overview is described below.

The docx parsing is based on ISO/IEC 29500-1:2016 [https://www.iso.org/standard/71691.html]
that describes the office open xml file formats. In a nutshell
a docx file is a zipped directory that contains a set of xml files.

The folder/file structured expected when unzipping a .docx file is:
```
        /_rels/
            .rels
            [Content_Types].xml
            ...
        /word/
            document.xml
            /_rels/
                document.xml.rels
            ...
```

The two xmls that will use to extract translatable content are:
* /word/document.xml -> Contains all(*) the textual content of the document
* /word/_rels/document.xml.rels -> Contains relationships details linked
  from /word/document.xml. These relationships will be used to extract
  external hyperlinks that are associated with text elements in the document

The string segmentation follows the segmentation of the textual content
into pararagraphs as this is defined inside document.xml.

A simplified version of the two files:
* word/document.xml
```
<w:document>
    <w:body>
        <w:p>
            ...
            <w:r>
                ...
                <w:t xml:space="preserve">hello</w:t>
                ...
            </w:r>
            ...
            <w:r>
                ...
                <w:t xml:space="preserve">world</w:t>
                ...
            </w:r>
            ...
        </w:p>
        ...
	<w:p>
            <w:hyperlink r:id="rId6">
                ...
                <w:r>
                     ...
                     <w:t xml:space="preserve">this is a link</w:t>
                     ...
                </w:r>
                ...
            </w:hyperlink>
        </w:p>
    </w:body>
</w:document>
```

* word/_rels/document.xml.rels
```
<Relationships>
    ...
    <Relationship Id="rId6" Target="https://www.transifex.com/" TargetMode="External"/>
    ...
</Relationships>
```

Parsing
-------
* All paragraphs (w:p elements) are inspected
* For each paragraph all child text elements (w:t) are combined
  into a single source string (seperated by pseudo tags <tx>)
  - For text elements surrounded by a hyperlink (w:hyperlink) the
    pseudo tag also includes the href extracted from `document.xml.rels`
* In each paragraph we add a special attribute (`txid`) which contains
  the hash of the final extracted source string. This will be helpful
  when compiling back the .docx file.

Complilation
------------
* All paragraphs (w:p elements) are inspected
* For each paragraph extract the attribute (`txid`) and find the
  corresponding translation by hash from the stringset provided
* Split the translation into parts by separating it at pseudo tags (<tx>)
* For each child text element (w:t) in the paragraph replace the text
  with the appropriate part of the translation

Tricks:
* There are some text elements that contain only spaces. In order to
  reduce noise from the extracted source string, we ignore them
  and prepend additional whitespace to the next text element.
  The reverse process is also taking place on complilation where
  we detect these text elements and we remove trailing spaces
  from the parts that need to.

* When the number of the text elements in a paragraph are not equal
  to the translation parts of the translation, we either:
  * remove text elements from the xml
  * combine multiple translation parts into one

How to test
-----------
Use as a guide the unit tests in order
to check how parsing and compilation behaves.

Reviewer checklist
------------------

Code:

PR:
